src/snowflake/snowpark/dataframe.py

```diff
@@ -343,6 +349,143 @@ def _disambiguate(
+            curr.append(DataFrameLineageNode(batch_id, stmt_cache))
+            if batch_id not in visited_batch_id:
+                visited_batch_id.add(batch_id)
+
+    lineage = []
+
+    while curr:
+        next: List[DataFrameLineageNode] = []
+        for node in curr:
+            # tracing updates
+            source_id = node.get_source_id()
+            if source_id not in visited_source_id:
+                visited_source_id.add(source_id)
+                lineage.append(node)
+
+            # explore next layer
+            for child_id in node.children:
+                if child_id in visited_batch_id:
+                    continue
+                visited_batch_id.add(child_id)
+                next.append(DataFrameLineageNode(child_id, node.stmt_cache))
+
+        curr = next
+
+    return lineage
+
+
+def dataframe_exception_handler(func):
+    """
+    Decorator to handle exceptions in DataFrame operations. Currently, it
+    only handles SnowparkSQLException. It captures the original stack trace and
+    adds additional debug information to the exception message, including
+    the traceback of the dataframe operations that could have caused the error.
+
+    We re-raise the original exception if:
+    1. The exception is already handled (i.e., the exception handler has been
+         invoked multiple times in case of a nested dataframe operation).
+    2. The exception is not a SnowparkSQLException.
+    3. We cannot infer the lineage of the dataframes involved in the exception.
+    """
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except SnowparkSQLException as original_exception:
+            # capture original stack trace
+            error_type, error_value, tb = sys.exc_info()
+            formatted_last_frame = str(
+                traceback.format_exception(error_type, error_value, tb)[-1]
+            )
+
+            if "--- Additional Debug Information ---" in formatted_last_frame:
+                # The exception handler can be invoked multiple times in
+                # case of a nested dataframe operation. We only want to
+                # add the debug information once. If the exception is
+                # already handled, we just re-raise it.
+                raise original_exception
+
+            lineage_trace_len = 0
+            try:
+                # get the dataframe lineage
+                dataframes_involved = []
+                for arg in args:
+                    if isinstance(arg, DataFrame):
+                        dataframes_involved.append(arg)
+
+                df_lineage = _get_df_lineage(dataframes_involved)
+                lineage_trace_len = len(df_lineage)
+                show_lineage_len = int(
+                    os.environ.get(
+                        "SNOWPARK_PYTHON_DATAFRAME_LINEAGE_LENGTH_ON_ERROR", 5
+                    )
+                )
```

Comment:
in the doc you mentioned this is turned off by default, but from the code I think it's turned on by default?

Reply:
I don't understand what you mean by "turned off by default". Do you mean the process of us adding debug logs or the environment variable to get the df transformation trace length?

The additional debug logs is turned off by default because ast_id is None when AST is turned off. And when ast_id for dataframe is None, we cannot collect any trace. Which is why this feature is turned off by default. Once AST is enabled, we also enable the additional debug logs.

The environment variable can be used to control the trace length. The default trace length is 5.